roaring/roaring.go

```diff
@@ -395,32 +399,282 @@ func (b *Bitmap) Intersect(other *Bitmap) *Bitmap {
 	iiter, _ := b.Containers.Iterator(0)
 	jiter, _ := other.Containers.Iterator(0)
 	i, j := iiter.Next(), jiter.Next()
 	ki, ci := iiter.Value()
 	kj, cj := jiter.Value()
 	for i || j {
 		if i && (!j || ki < kj) {
-			output.Containers.Put(ki, ci.Clone())
+			target.Containers.Put(ki, ci.Clone())
 			i = iiter.Next()
 			ki, ci = iiter.Value()
 		} else if j && (!i || ki > kj) {
-			output.Containers.Put(kj, cj.Clone())
+			target.Containers.Put(kj, cj.Clone())
 			j = jiter.Next()
 			kj, cj = jiter.Value()
 		} else { // ki == kj
-			output.Containers.Put(ki, union(ci, cj))
+			target.Containers.Put(ki, union(ci, cj))
 			i, j = iiter.Next(), jiter.Next()
 			ki, ci = iiter.Value()
 			kj, cj = jiter.Value()
 		}
 	}
-	return output
+}
+
+// unionIntoTarget stores the union of b and others into target. b and others will
+// be left unchanged (unless one of them is also target), but target will be modified
+// in place.
+//
+// This function performs an n-way union of n bitmaps. It performs this in an
+// optimized manner looping through all the bitmaps and performing unions one
+// container at a time. As a result, instead of generating many intermediary
+// containers for each union operation for a given container key, only one
+// new container needs to be allocated (or re-used) regardless of how many bitmaps
+// participate in the union. This significantly reduces allocations. In addition,
+// because we perform the unions one container at a time across all the bitmaps, we
+// can calculate summary statistics that allow us to make more efficient decisions
+// up front. For example, imagine trying to perform a union across the following three
+// bitsets:
+//
+//     1. Bitmap A: Single array container at key 0 with 400 values in it.
+//     2. Bitmap B: Single array container at key 0 with 500 values in it.
+//     3. Bitmap C: Single array container at key 0 with 3500 values in it.
+//
+// Naive approach:
+//
+//     1. Perform union of bitmap A and B, container by container
+//         a. 400 + 500 < ArrayMaxSize so likely we will choose to allocate a new array
+//            container and then perform a unionArrayArray operation to merge the two
+//            arrays into the new array container.
+//     2. Perform a union of the bitmap generated in the step above with bitmap C.
+//        900 + 3500 > ArrayMaxSize so we will need to upgrade to a bitset container which
+//        we will have to allocate, and then we will have to perform two unions into the
+//        new bitmap container: one for the array container generated in the previous step,
+//        and one for the bitset container in bitmap C.
+//
+// Approach taken by this function:
+//
+//     1. Detect that bitmaps A, B, and C all have containers for key 0.
+//     2. Estimate the resulting cardinality of the union of all their containers to be
+//        400 + 500 + 3500 > ArrayMaxSize and decide upfront to use a bitset for the target
```

Comment:
I don't think this math is necessarily right -- we don't know whether the values overlap, so we don't know how many values we'll end up with. Not sure it's worth worrying about, though.

Reply:
yeah its definetly just a rough approximation. It prevents us from having to use a bitmap everytime, but will definitely be biased towards being more aggressive and using a bitmap because we can't be sure.

Reply:
I tried to convey that by using the word "estimate" but maybe I can make the comment more explicit. I just didn't want to use a bitmap if I was unioning two containers, each with a cardinality of 1

Reply:
Updated the comment!